(vim) Add vim

matsuhav opened the pull request:
<!-- Provide a general summary of your changes in the Title above, prefixed with (packageName) -->

## Description
Migrated from ferventcoder/chocolatey-packages.
Changed package source to vim/vim-win32-installer and updated nuspec.
Now package is obtained as zip to provide installation options.

## Motivation and Context
Vim package is outdated because package source's project(unofficial) is inactive.
There is an official nightly vim build from vim.org [vim-win32-installer](https://github.com/vim/vim-win32-installer).
And there are no packages to install official builds on chocolatey.
I wrote a AU script to update package, but original repository [ferventcoder/chocolatey-packages](https://github.com/ferventcoder/chocolatey-packages) seems to not support AU.
And I want to migrate vim package and update it.

## How Has this Been Tested?
I used chocolatey-test-environment to test.
Installation, uninstallation, with/without parameters
<!-- Include details of your testing environment, and the tests you ran to -->
`choco install -s . vim -dv`
`choco uninstall -s . vim -dv`
`choco install -s . vim --params '/NoDesktopShortcuts'` and other params combinations.
<!-- see how your change affects other areas of the script, etc. -->

I changed to install from zip file to provide installation parameters. Due to this change, I changed nuspec and install/uninstall scripts and added beforemodify script. 
At first I submitted as a different package to community and it's [here](https://chocolatey.org/packages/vim-latest/). It passed automated validations. I will delete this after this PR closed.

## Screenshot (if appropriate, usually isn't needed):

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Migrated package (a package has been migrated from another repository)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey/chocolatey-coreteampackages/blob/master/CONTRIBUTING.md)
- [x] The added/modified package passed install/uninstall in the chocolatey test environment.
- [x] The changes only affect a single package (not including meta package).

<!-- The following section can be removed if the package has not been migrated from another location -->
## Original Location
- [ferventcoder/chocolatey-packages](https://github.com/ferventcoder/chocolatey-packages)
- Open Issues
  - [x] another PR to change about the vim package https://github.com/ferventcoder/chocolatey-packages/pull/304
- [x] The opened PR that removes the package from the original location https://github.com/ferventcoder/chocolatey-packages/pull/306
- [x] The [migration guidelines](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Package-migration-process) have been followed

majkinetor commented on automatic/vim/README.md:
```diff
@@ -0,0 +1,19 @@
+
+The software is installed to chocolatey tools directory.
+
+#### Parameters
```
Its named `Package parameters`. You miss `Features` too.

Also, I suggest in notes to specify relation to other packages and why this one differes. (like the one you gave on gallery).

majkinetor commented on automatic/vim/tools/chocolateybeforemodify.ps1:
```diff
@@ -0,0 +1,2 @@
+﻿Write-Debug 'Restarting explorer.'
```
You restart explorer every time on update while it might not be necessary (particularly given that this is pretty intrusive).

majkinetor commented on automatic/vim/tools/chocolateyinstall.ps1:
```diff
@@ -0,0 +1,50 @@
+$packageArgs = @{
+        packageName    = $env:ChocolateyPackageName
+        unzipLocation  = $toolsDir
+        url            = 'https://github.com/vim/vim-win32-installer/releases/download/v8.1.2056/gvim_8.1.2056_x86.zip'
```
You should embedd vim zip into the package. It has permissive license and is not big.

majkinetor commented on automatic/vim/tools/chocolateyinstall.ps1:
```diff
@@ -0,0 +1,50 @@
+$installpopup = '-install-popup'
+$installicons = '-install-icons'
+# restart exeplorer.  From vim-tux.install
+if ($params['RestartExplorer'] -eq 'true') {
```
You should tidy up installer by moving some boilerplate into helpers.ps1

majkinetor commented on automatic/vim/vim.nuspec:
```diff
@@ -0,0 +1,40 @@
+    <releaseNotes></releaseNotes>
+    <projectUrl>https://www.vim.org</projectUrl>
+    <packageSourceUrl>https://github.com/chocolatey-community/chocolatey-coreteampackages/</packageSourceUrl>
+    <tags>vim editor vi gvim admin vim-win32-installer</tags>
```
missing `cross-platform` `foss` obligatory tags

matsuhav commented:
1. In README.md,
    1. Renamed to `Package parameters`.
    2. Added `Features`.
    3. Added `Notes`.
2. Changed chocolateybeforemodify.ps1 to restart explorer only when needed.
3. Changed to embed zip and added LICENSE.txt and VERIFICATION.txt too.
4. Made helpers.ps1 and cleaned up chocolateyinstall.ps1
5. Updated nuspec.

This package's original location is creator of chocolatey and I completely forgot about README.md and nuspec, sorry. It's good opportunity to update them.

majkinetor commented:
BTW, missed you need to add `legal` folder in nuspec

majkinetor commented:
/cc @AdmiringWorm @gep13  Needs chocolatey user.

matsuhav commented:
Thank you for reviews!
I added `InstallDir` option and made that default value to `Get-ToolsLocation`.
The params section on README.md is also updated.

4 Write-Debug is now smart one line.

majkinetor commented:
Some small problems now:

- You dodn't need `set-noshim` any more as choco doesn't automatically add shims outside of its folder.
- However, vim and givm are not available on the path so they are not easy to start. You should now add those shims manually in install and remove them in uninstall.

matsuhav commented:
I removed no-shim things.

Install script runs vim's install.exe.  It creates bat files on `$Windows` directory (normally `C:\Windows`), which is on PATH. These bat files does more things than shim and you can invoke vim/gvim from terminal. They are removed when uninstall script calls uninstal.exe.
I implemented them already.

Other 2 commits are from my careless miss.

majkinetor commented:
Ok this can be merged once chocolatey user is added.

AdmiringWorm commented:
@majkinetor chocolatey user have been added.

@matsuhav if you haven't done so already, please open a PR to ferventcoders repository to remove all of the package files from there (except the icon).

majkinetor commented:
https://github.com/ferventcoder/chocolatey-packages/pull/306

majkinetor commented:
@matsuhav for some reason package uninstall is failing on AV ? Can you take a look into that before we merge this ?

matsuhav commented:
The uninstaller was renamed [a few days ago](https://github.com/vim/vim/commit/30e8e73506e4522ef4aebf7d525c0e6ffe8805fd) for the first time more than 15 years.
It was perfect timing.

majkinetor commented:
> The uninstaller was renamed a few days ago for the first time more than 15 years.
It was perfect timing.

Amazing. 


Thanks for your work here.
